File: src/android/java/SystemUtilPlugin.java
package cordova.plugins;

import android.content.Context;
import android.text.TextUtils;
import android.util.Log;
import com.plugin.keystore.KeyStore;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.device.Device;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class SystemUtilPlugin extends CordovaPlugin {

  private static final String TAG = "SystemUtilPlugin";
  private String key = "";
  private String iv = "";
  private static String t = "";
  private static String p = "";
  private static String s = "";
  private static String decryptedS = ""; // 新增：存储解密后的s值
  
  @Override
  protected void pluginInitialize() {
    super.pluginInitialize();
    
    try {
      // 初始化 KeyStore，这会自动加载 keystore.so
      KeyStore keyStore = new KeyStore();
      
      // 从 native 库获取所有密钥
      key = keyStore.getAESKey();
      iv = keyStore.getAESIV();
      s = keyStore.getStringS();
      
      // 在初始化时直接解密s，后面直接使用decryptedS
      if (!TextUtils.isEmpty(s) && !TextUtils.isEmpty(key) && !TextUtils.isEmpty(iv)) {
        try {
          decryptedS = AESUtil.decryptCBC(s, key, iv);
          Log.d(TAG, "Native KeyStore initialized successfully, s decrypted");
        } catch (Exception e) {
          Log.e(TAG, "Failed to decrypt s: " + e.getMessage());
          decryptedS = "";
        }
      } else {
        Log.e(TAG, "Key, iv or s is empty, cannot decrypt s");
      }
      
      Log.d(TAG, "Native KeyStore initialized successfully");
      
    } catch (Exception e) {
      Log.e(TAG, "Failed to initialize native KeyStore: " + e.getMessage());
    }
  }

  @Override
  public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
    if (action.equals("putIn")) {
      return putIn(args, callbackContext);
    } else if (action.equals("putPub")) {
      return putPub(args, callbackContext);
    } else if (action.equals("putOut")) {
      return putOut(args, callbackContext);
    } else {
      callbackContext.error("未知的action: " + action);
      return false;
    }
  }

  private boolean putIn(JSONArray args, CallbackContext callbackContext) {
    try {
      JSONObject obj = args.getJSONObject(0);
      t = obj.optString("accessToken", "");
      p = obj.optString("prvKey", "");
      callbackContext.success();
    } catch (JSONException e) {
      callbackContext.error("缓存数字信封失败");
    }
    return true;
  }

  private boolean putPub(JSONArray args, CallbackContext callbackContext) {
    try {
      Context context = cordova.getActivity().getApplicationContext();
      JSONObject obj = args.getJSONObject(0);
      String fingerKey = obj.optString("fingerKey", "");
      String faceKey = obj.optString("faceKey", "");
      
      if (!TextUtils.isEmpty(fingerKey)) {
        String encryptedPubKey = AESUtil.encryptCBC(fingerKey, key, iv);
        SharedPrefsUtil.savePreference(context, "fin2Key", encryptedPubKey);
      }
      if (!TextUtils.isEmpty(faceKey)) {
        String encryptedPubKey = AESUtil.encryptCBC(faceKey, key, iv);
        SharedPrefsUtil.savePreference(context, "fac2Key", encryptedPubKey);
      }
      callbackContext.success();
    } catch (Exception e) {
      callbackContext.error("缓存指纹公钥失败");
    }
    return true;
  }

  private boolean putOut(JSONArray args, CallbackContext callbackContext) {
    try {
      JSONObject obj = args.getJSONObject(0);
      String name = obj.optString("name", "");
      
      switch (name) {
        case "t":
          callbackContext.success(t);
          break;
        case "p":
          callbackContext.success(p);
          break;
        case "s":
          // 直接返回已经解密的s值
          callbackContext.success(decryptedS);
          break;
        case "all":
          JSONObject all = new JSONObject();
          all.put("t", t);
          all.put("p", p);
          // 直接使用已经解密的s值
          all.put("s", decryptedS);
          callbackContext.success(all);
          break;
        case "sign":
          JSONObject result = new JSONObject();
          String a1 = obj.optString("a1", "");
          String a2 = obj.optString("a2", "");
          String a3 = obj.optString("a3", "");
          
          // 1. 使用SM3加密(a1 + a2 + a3 + t)
          String m1 = a1 + a2 + a3;
          if (!TextUtils.isEmpty(t)) {
            m1 += t;
          }
          String sm3Hash = SM3Util.sm3(m1);
          
          // 2. 获取设备UUID
          String deviceUuid = Device.uuid != null ? Device.uuid : "";
          
          // 3. 直接使用已经解密的s值作为SM2公钥
          String sm2PublicKey = decryptedS;
          
          // 4. 使用SM2加密(sm3Hash + deviceUuid)
          String toBeEncrypted = sm3Hash + deviceUuid;
          String encryptedData = "";
          if (!TextUtils.isEmpty(sm2PublicKey) && !TextUtils.isEmpty(toBeEncrypted)) {
            encryptedData = SM2Util.encryptWithSM2(toBeEncrypted, sm2PublicKey);
          }
          
          result.put("b", encryptedData);
          callbackContext.success(result);
          break;
        default:
          callbackContext.error("参数错误");
          break;
      }
    } catch (Exception e) {
      callbackContext.error(e.getMessage());
    }
    return true;
  }
}

File: src/ios/oc/SystemUtilPlugin.m
#import "SystemUtilPlugin.h"
#import "AESUtil.h"
#import "MD5Util.h"
#import "RSAUtil.h"
#import "StorageUtil.h"
#import "SM3Util.h"
#import "SM2Util.h"
#import <UIKit/UIKit.h>

@implementation SystemUtilPlugin {
    NSString *_accessToken;
    NSString *_prvKey;
    NSString *_decryptedS; // 存储解密后的s值
}

static NSString *const kKey = @""; // 这里应该是实际的密钥
static NSString *const kIV = @"";  // 这里应该是实际的IV
static NSString *const kEncryptedS = @""; // 这里应该是加密的s值
static NSString *const kEncryptedFS = @"";
static NSString *const kEncryptedPS = @"";

#pragma mark - 初始化方法

- (void)pluginInitialize {
    [super pluginInitialize];
    
    // 在初始化时直接解密s，后面直接使用_decryptedS
    if (kEncryptedS.length > 0 && kKey.length > 0 && kIV.length > 0) {
        @try {
            _decryptedS = [AESUtil decryptCBC:kEncryptedS key:kKey iv:kIV];
            NSLog(@"SystemUtilPlugin: s解密成功，长度: %lu", (unsigned long)_decryptedS.length);
        } @catch (NSException *exception) {
            NSLog(@"SystemUtilPlugin: s解密失败: %@", exception.reason);
            _decryptedS = @"";
        }
    } else {
        NSLog(@"SystemUtilPlugin: kEncryptedS、kKey或kIV为空，无法解密s");
    }
    
    [StorageUtil savePreference:@"fin1Key" value:kEncryptedFS];
    [StorageUtil savePreference:@"fac1Key" value:kEncryptedPS];
}

#pragma mark - 主要方法

- (void)putIn:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    
    @try {
        NSDictionary* params = [command.arguments objectAtIndex:0];
        
        if (params) {
            _accessToken = [params objectForKey:@"accessToken"] ?: @"";
            _prvKey = [params objectForKey:@"prvKey"] ?: @"";
            
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                              messageAsString:@"参数错误"];
        }
    } @catch (NSException *exception) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                          messageAsString:@"缓存数字信封失败"];
    }
    
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)putPub:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    
    @try {
        NSDictionary* params = [command.arguments objectAtIndex:0];
        NSString* fingerKey = [params objectForKey:@"fingerKey"] ?: @"";
        NSString* faceKey = [params objectForKey:@"faceKey"] ?: @"";
        
        if (fingerKey.length > 0) {
            NSString* encryptedPubKey = [AESUtil encryptCBC:fingerKey key:kKey iv:kIV];
            [StorageUtil savePreference:@"fin2Key" value:encryptedPubKey];
            
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        } else if (faceKey.length > 0) {
            NSString* encryptedPubKey = [AESUtil encryptCBC:faceKey key:kKey iv:kIV];
            [StorageUtil savePreference:@"fac2Key" value:encryptedPubKey];
            
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                              messageAsString:@"公钥不能为空"];
        }
    } @catch (NSException *exception) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                          messageAsString:@"缓存指纹公钥失败"];
    }
    
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)putOut:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    
    @try {
        NSDictionary* params = [command.arguments objectAtIndex:0];
        NSString* name = [params objectForKey:@"name"] ?: @"";
        
        if ([name isEqualToString:@"t"]) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK 
                                             messageAsString:_accessToken ?: @""];
        } else if ([name isEqualToString:@"p"]) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK 
                                             messageAsString:_prvKey ?: @""];
        } else if ([name isEqualToString:@"s"]) {
            // 直接返回已经解密的s值
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK 
                                             messageAsString:_decryptedS ?: @""];
        } else if ([name isEqualToString:@"all"]) {
            NSMutableDictionary* resultDict = [NSMutableDictionary dictionary];
            
            resultDict[@"t"] = _accessToken ?: @"";
            resultDict[@"p"] = _prvKey ?: @"";
            // 直接使用已经解密的s值
            resultDict[@"s"] = _decryptedS ?: @"";
            
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK 
                                         messageAsDictionary:resultDict];
        } else if ([name isEqualToString:@"sign"]) {
            NSString* a1 = [params objectForKey:@"a1"] ?: @"";
            NSString* a2 = [params objectForKey:@"a2"] ?: @"";
            NSString* a3 = [params objectForKey:@"a3"] ?: @"";
            
            // 1. 使用SM3加密(a1 + a2 + a3 + t)
            NSString* m1 = [NSString stringWithFormat:@"%@%@%@", a1, a2, a3];
            
            if (_accessToken.length > 0) {
                m1 = [m1 stringByAppendingString:_accessToken];
            }
            
            NSLog(@"SM3输入: %@", m1);
            NSString* sm3Hash = [SM3Util sm3:m1];
            NSLog(@"SM3输出: %@", sm3Hash);
            
            // 2. 获取设备UUID
            NSString* deviceUUID = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
            NSLog(@"设备UUID: %@", deviceUUID);
            
            // 3. 直接使用已经解密的s值作为SM2公钥
            NSString* sm2PublicKey = _decryptedS;
            NSLog(@"SM2公钥长度: %lu", (unsigned long)sm2PublicKey.length);
            
            // 4. 使用SM2加密(sm3Hash + deviceUuid)
            NSString* toBeEncrypted = [NSString stringWithFormat:@"%@%@", sm3Hash ?: @"", deviceUUID ?: @""];
            NSLog(@"SM2待加密数据: %@", toBeEncrypted);
            
            NSString* encryptedData = @"";
            
            if (sm2PublicKey.length > 0 && toBeEncrypted.length > 0) {
                encryptedData = [SM2Util encryptWithSM2:toBeEncrypted publicKeyStr:sm2PublicKey];
                NSLog(@"SM2加密结果长度: %lu", (unsigned long)encryptedData.length);
            }
            
            NSMutableDictionary* resultDict = [NSMutableDictionary dictionary];
            resultDict[@"b"] = encryptedData ?: @"";
            
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK 
                                         messageAsDictionary:resultDict];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                              messageAsString:@"参数错误"];
        }
    } @catch (NSException *exception) {
        NSLog(@"SystemUtilPlugin异常: %@", exception);
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                          messageAsString:exception.reason ?: @"获取失败"];
    }
    
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

@end